### ISSUE TITLE
Handling assignments to list items has some limitations

### ISSUE THREAD

**Issue by ptomulik** [vimeo/psalm#9484]
In short, psalm reports errors for the following examples, which are correct in fact:

https://psalm.dev/r/8972d76050 (copying a list to another list).
https://psalm.dev/r/74126f725d (appending to list end, but using an offset)
https://psalm.dev/r/fe3d1f172a (altering elements within a loop over continuous range of indices starting from 0)

**Comment by othercorey** [comment]
I the first example, psalm would have to track that none of the indexes were skipped in order to assert that it returns a list again.

It seems simple in that example, but imagine other conditions inside the foreach.

**Comment by ptomulik** [comment]
I agree that this may be very complicated. Honestly, I created this issue mainly for the purpose of being referenced from other issues and PRs.
